### PR TITLE
node/objects: Distribute objects in container via replication protocol 

### PR DESCRIPTION
**Pull request by cthulhu-rider**
* based on and blocked by #2801 (first 2 commits)

### CODE REVIEW

**Comment by roman-khimov** [comment]
post-0.41.0

File: pkg/services/object/put/distributed.go [hunk 1]
```diff
@@ -115,25 +123,49 @@ func (x errIncompletePut) Error() string {
...
+			t.encodedObject, err = encodeReplicateRequestWithoutPayload(t.localNodeSigner, *hdr, int(payloadLen))
+		}
+		if err != nil {
+			return fmt.Errorf("encode object into binary: %w", err)
+		}
+	} else if payloadLen > 0 {
+		b := getPayload()
+		if cap(b) < int(payloadLen) {
+			putPayload(b)
+			b = make([]byte, 0, payloadLen)
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/services/object/put/distributed.go:146]: `payloadLen <= MaxObjectSize` here. In practice, the limit wont be so big. I can comment this line about it

File: pkg/services/object/put/distributed.go [hunk 2]
```diff
@@ -28,7 +30,13 @@ type distributedTarget struct {
 	obj     *objectSDK.Object
 	objMeta object.ContentMeta
 
-	payload []byte
+	localOnly            bool
+	localNodeInContainer bool
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/services/object/put/distributed.go:34]: missed `nodeInPlacement` came after rebase. It's no longer needed with new changes

File: cmd/neofs-node/transport.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	apistatus "github.com/nspcc-dev/neofs-sdk-go/client/status"
+)
+
+type transport struct {
+	clients *coreClientConstructor
+}
+
+// SendReplicationRequestToNode connects to described node and sends prepared
+// replication request message to it.
+func (x *transport) SendReplicationRequestToNode(ctx context.Context, req []byte, node coreclient.NodeInfo) error {
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply cmd/neofs-node/transport.go:23]: nope for the sake of performance: node sends same object to several nodes

File: pkg/services/object/put/proto.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+func encodeReplicateRequestWithoutPayload(signer neofscrypto.Signer, hdr object.Object, pldLen int) (encodedObject, error) {
+	var res encodedObject
+	id, ok := hdr.ID()
+	if !ok {
+		return res, errors.New("missing object ID")
+	}
+
+	sig, err := signer.Sign(id[:])
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/services/object/put/proto.go:76]: i dont think `Encode` will survive until SDK release

**Review by carpawell — APPROVED** [review]
I think, we need a deep refactor in the distribution code, it went through a lot of changes. We have many "dynamic" code (where `[]byte` may mean different things, `obj` and `encodedObject` are neighbors fields, etc), traverser is not the only one who knows about object placement, some TODOs and proto code are also here but we need to go further.

**Pull request merged**